typing/typetexp.ml

```diff
@@ -50,8 +51,209 @@ type error =
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
-(** Map indexed by type variable names. *)
-module TyVarMap = Misc.Stdlib.String.Map
+module TyVarEnv : sig
+  val reset : unit -> unit
+  (* see mli file *)
+
+  val is_in_scope : string -> bool
+
+  val add : string -> type_expr -> unit
+  (* add a global type variable to the environment *)
+
+  val with_local_scope : (unit -> 'a) -> 'a
+  (* see mli file *)
+
+  type poly_univars
+  val with_univars : poly_univars -> (unit -> 'a) -> 'a
+  (* evaluate with a locally extended set of univars *)
+
+  val make_poly_univars : string list -> poly_univars
+  (* see mli file *)
+
+  val check_poly_univars : Env.t -> Location.t -> poly_univars -> type_expr list
+  (* see mli file *)
+
+  type policy
+  val fixed_policy : policy (* no wildcards allowed *)
+  val extensible_policy : policy (* common case *)
+  val univars_policy : policy (* fresh variables are univars (in methods) *)
```

Comment:
You decomposed `policy` in two independent dimensions. For later: should we consider an API that expose the two dimensions, instead of hiding them as an opaque type? This could clarify the meaning of the policies, and it might even simplify/clarify the caller-side code.

Reply:
I waffled on this quite a bit. I'm not particularly happy with where this settled, but it seemed better (i.e. more concise) this way. The opacity is because `Universal`/`Fixed` never happens. Would be happy for someone else to have a go at a better design or for a concrete suggestion of a redesign here.

Reply:
I am not completely convinced that keeping the three variants would not be clearer, maybe renaming them to:
```ocaml
type flavor = Closed | Unification | Univars
```
The issue that I have with the `fixed/extensible` axis is that the `Fixed`  (renamed to `Closed` above) policy is nowadays only ever used to fail early on the presence of a wildcard in a type definition. Typically, we could remove the `Fixed`  policy completely and accept few more strange type declarations like:
```
type t = (int as 'a) * ( _ as 'a)
```
which is currently rejected as type declaration, whereas
```ocaml
type t = (int as 'a) * ( 'any_fresh_name as 'a)
```
is accepted.

I am not really arguing for the change, but I wonder if this distinction warrants the creation of an independent new axis in the policy type?
In term of concision, I have the impression that the variant version is as short or shorter (https://github.com/Octachron/ocaml/commit/5c9e9e1bb341ad6d4e17771a349da568ee06c23f)?

Reply:
That's a good point about `Closed`. My goal was to not change behavior, and so I didn't even really consider this idea. But I agree that this seems a poor reason for complexity. I'm very happy to adopt the change back to a single variant -- or just to a variant with only two constructors. I'm the new contributor here, so just tell me what to do and I'll do it. :)

Reply:
After looking at `globalize_used_variables`, I think it makes sense to keep the ternary policy in order to avoid a gimbal lock in  `globalize_used_variables`.  I also tend to prefer to avoid splitting this type in a product because I see it as an answer to the single question of `how should free type variables in the type expression be interpreted?` but that's debatable.